### PR TITLE
Create Event lifecycle test - +5 endpoint coverage

### LINKED ISSUE #86288: Promote CoreV1Event test - +5 coverage
- [x] [mock test ticket](#)
- [x] [PR](https://github.com/kubernetes/kubernetes/pull/86858)
- [ ] [promotion](https://github.com/kubernetes/kubernetes/pull/89753)

This issue proposes the writing of a new test for the API endponts `readCoreV1NamespacedEvent`, `patchCoreV1NamespacedEvent`, `listCoreV1NamespacedEvent`, `listCoreV1EventForAllNamespaces`, `deleteCoreV1NamespacedEvent`, and `createCoreV1NamespacedEvent` with the intention of promoting it to Conformance. This issue contains a very basic mock test in order to start the conversation about what a good e2e test would look like for this endpoint. The code below is verified to be hitting the intended endpoint, show as per the queries in APIsnoop's live view of the cluster's audit logs.

## Deploy APISnoop

```shell
kubectl apply -f ~/cncf/apisnoop/deployment/k8s/raiinbow.yaml
```

# Identifying an untested feature using APISnoop

According to these query results from APIsnoop, we are lacking in tests for the following endpoints related to Events

```sql-mode
SELECT
  operation_id,
  -- k8s_action,
  -- path,
  description
  FROM untested_stable_core_endpoints
  where path not like '%volume%'
  and operation_id like '%Event%'
 ORDER BY operation_id desc
 -- LIMIT 25
       ;
```

```sql-mode
             operation_id              |             description              
---------------------------------------|--------------------------------------
 replaceCoreV1NamespacedEvent          | replace the specified Event
 readCoreV1NamespacedEvent             | read the specified Event
 patchCoreV1NamespacedEvent            | partially update the specified Event
 listCoreV1NamespacedEvent             | list or watch objects of kind Event
 listCoreV1EventForAllNamespaces       | list or watch objects of kind Event
 deleteCoreV1NamespacedEvent           | delete an Event
 deleteCoreV1CollectionNamespacedEvent | delete collection of Event
 createCoreV1NamespacedEvent           | create an Event
(8 rows)

```

We chose the following 5 endpoints for this **mock** test:
- `readCoreV1NamespacedEvent`
- `patchCoreV1NamespacedEvent`
- `listCoreV1NamespacedEvent`
- `listCoreV1EventForAllNamespaces`
- `deleteCoreV1NamespacedEvent`
- `createCoreV1NamespacedEvent`

# API References

-   [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
-   [client-go - Events](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/event.go)

# The mock test

## Example in Go

Test outline:
1. create Event in namespace default with custom fields

2. list all events in all namespaces
   find event created in #1 in response of #2

3. update the content of the event created in #1

4. get events by name in namespace default
   ensure that the event updated from #4

5. delete event by name

6. list all events in namespace
   ensure event in namespace default was deleted

```shell
go get -v -u k8s.io/apimachinery/pkg/apis/meta/v1
go get -v -u k8s.io/client-go/kubernetes
go get -v -u k8s.io/client-go/tools/clientcmd
```

This mock test uses client-go's standard functions for talking to the Event area of the API.

```go
package main

import (
  "fmt"
  "flag"
  "os"
  //"encoding/json"
  v1 "k8s.io/api/core/v1"
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  "k8s.io/client-go/tools/clientcmd"
  "k8s.io/apimachinery/pkg/types"
)

func main() {
  // uses the current context in kubeconfig
  kubeconfig := flag.String("kubeconfig", fmt.Sprintf("%v/%v/%v", os.Getenv("HOME"), ".kube", "config"), "(optional) absolute path to the kubeconfig file")
     flag.Parse()
  config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
  if err != nil {
    fmt.Println(err)
    return
  }
  // make our work easier to find in the audit_event queries
  config.UserAgent = "live-test-writing"
  // creates the clientset
  clientset, _ := kubernetes.NewForConfig(config)

  eventTestName := "event-test"

  // create a test event in default namespace
  _, err = clientset.CoreV1().Events("default").Create(&v1.Event{
    ObjectMeta: metav1.ObjectMeta{
      Name: eventTestName,
    },
    Message: "This is a test event",
    Reason: "Test",
    Type: "Normal",
    Count: 1,
  })
  if err != nil {
    fmt.Println("[error]", err)
    return
  }
  fmt.Println("[status] test event created")

  // get a list of Events in all namespaces
  eventsList, err := clientset.CoreV1().Events("").List(metav1.ListOptions{})
  if err != nil {
    fmt.Println("[error]", err)
    return
  }
  if len(eventsList.Items) == 0 {
    fmt.Println("[error] no events found")
    return
  }
  foundCreatedEvent := false
  var eventCreatedName string
  for _, val := range eventsList.Items {
    if val.ObjectMeta.Name == eventTestName && val.Message == "This is a test event" {
      foundCreatedEvent = true
      eventCreatedName = val.ObjectMeta.Name
    }
  }
  if foundCreatedEvent == false {
    fmt.Println("[error] unable to find event")
    return
  }
  fmt.Println("[status] created event found")

  // patch the event's message
  eventPatch := fmt.Sprintf(`{"message":"This is a test event - patched"}`)
  _, err = clientset.CoreV1().Events("default").Patch(eventTestName, types.StrategicMergePatchType, []byte(eventPatch))
  if err != nil {
    fmt.Println("[error]", err)
    return
  }
  fmt.Println("[status] patched event successfully")

  // get event by name
  event, err := clientset.CoreV1().Events("default").Get(eventCreatedName, metav1.GetOptions{})
  if err != nil {
    fmt.Println("[error]", err)
    return
  }
  if event.Message != "This is a test event - patched" {
    fmt.Println("[error]", err)
    return
  }
  fmt.Println("[status] found event with patched message")

  // delete original event
  err = clientset.CoreV1().Events("default").Delete(eventCreatedName, &metav1.DeleteOptions{})
  if err != nil {
    fmt.Println("[error]", err)
    return
  }

  // get a list of Events default namespace
  eventsList, err = clientset.CoreV1().Events("").List(metav1.ListOptions{})
  if err != nil {
    fmt.Println("[error]", err)
    return
  }
  if len(eventsList.Items) == 0 {
    fmt.Println("[error] no events found")
    return
  }
  foundCreatedEvent = false
  for _, val := range eventsList.Items {
    if val.ObjectMeta.Name == eventTestName && val.Message == "This is a test event - patched" {
      foundCreatedEvent = true
      eventCreatedName = val.ObjectMeta.Name
    }
  }
  if foundCreatedEvent == true {
    fmt.Println("[error] found event")
    return
  }

  fmt.Println("[status] complete")
}
```

# Verify with APISnoop

Discover useragents
```sql-mode
select distinct useragent from audit_event where bucket='apisnoop' and useragent not like 'kube%' and useragent not like 'coredns%' and useragent not like 'kindnetd%';
```

```sql-mode
     useragent     
-------------------
 live-test-writing
(1 row)

```

Display endpoints hit by the test
```sql-mode
select * from endpoints_hit_by_new_test where useragent like 'live%';
```

```
     useragent     |          operation_id           | hit_by_ete | hit_by_new_test 
-------------------+---------------------------------+------------+-----------------
 live-test-writing | createCoreV1NamespacedEvent     |          0 |               2 
 live-test-writing | deleteCoreV1NamespacedEvent     |          0 |               2 
 live-test-writing | listCoreV1EventForAllNamespaces |          0 |               2 
 live-test-writing | patchCoreV1NamespacedEvent      |          0 |               2 
 live-test-writing | readCoreV1NamespacedEvent       |          0 |               1 
(5 rows)                                                                            
```

Show the change in coverage after the mock test
```sql-mode
select * from projected_change_in_coverage;
```

```sql-mode
   category    | total_endpoints | old_coverage | new_coverage | change_in_number 
---------------|-----------------|--------------|--------------|------------------
 test_coverage |             438 |          183 |          188 |                5
(1 row)

```

# Final notes
If a test with these calls gets merged in, **conformance will go up by 5 endpoints**.

# 
/sig testing
/sig architecture
/area conformance

### PR DESCRIPTION
**Pull request by BobyMCbobs**
- [x] [mock test ticket](https://github.com/kubernetes/kubernetes/issues/86288)
- [x] [PR](#)
- [ ] [promotion](https://github.com/kubernetes/kubernetes/pull/89753)

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR adds a test to test the following untested endpoints:
- readCoreV1NamespacedEvent,
- patchCoreV1NamespacedEvent
- listCoreV1EventForAllNamespaces
- deleteCoreV1NamespacedEvent
- createCoreV1NamespacedEvent

**Which issue(s) this PR fixes**:
Fixes #86288

**Special notes for your reviewer**:
Adds +5 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
NONE
```

/sig testing

### CODE REVIEW

**Comment by BobyMCbobs** [comment]
I've updated this test to comply with what was recommended by @liggitt on #86857.

The test seems to fail with the following error when using `f.Namespace.Name` instead of `"default"`:
```
FAIL: failed to create test event
Unexpected error:
    <*errors.StatusError | 0xc002e26640>: {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {
                SelfLink: "",
                ResourceVersion: "",
                Continue: "",
                RemainingItemCount: nil,
            },
            Status: "Failure",
            Message: "Event \"event-test\" is invalid: involvedObject.namespace: Invalid value: \"\": does not match event.namespace",
            Reason: "Invalid",
            Details: {
                Name: "event-test",
                Group: "",
                Kind: "Event",
                UID: "",
                Causes: [
                    {
                        Type: "FieldValueInvalid",
                        Message: "Invalid value: \"\": does not match event.namespace",
                        Field: "involvedObject.namespace",
                    },
                ],
                RetryAfterSeconds: 0,
            },
            Code: 422,
        },
    }
    Event "event-test" is invalid: involvedObject.namespace: Invalid value: "": does not match event.namespace
occurred
```

**Comment by liggitt** [comment]
you also need to create the object the event is for in the test namespace

**Comment by BobyMCbobs** [comment]
> you also need to create the object the event is for in the test namespace

Thank you @liggitt! Updated in 13f335c532b

File: test/e2e/common/events.go
Comment (spiffxp) [inline test/e2e/common/events.go]: I think this comment should make explicit why we're hitting all namespaces, otherwise a well meaning test maintainer may later decide to optimize and hit just `f.Namespace.Name`

File: test/e2e/common/events.go
Comment (spiffxp) [inline test/e2e/common/events.go]: look for it by name/namespace not name/data

**Review by spiffxp — COMMENTED** [review]
This seems a reasonable test but note that I'm not sure it's eligible for Conformance.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements says

>we make no guarantees about the contents of events, nor their delivery

I get that in this case the test itself is setting the content, so we're not reliant on arbitrary components' non-existent event contract, I'm still not sure we're able to guarantee that this synthetic event will be around even for the duration of the test.

What do others think?

File: test/e2e/common/events.go
Comment (BobyMCbobs) [reply test/e2e/common/events.go]: Updated in 2a535e4f38c

File: test/e2e/common/events.go
Comment (BobyMCbobs) [reply test/e2e/common/events.go]: Updated in d8909147c30

File: test/e2e/framework/events/events.go
Comment (spiffxp) [inline test/e2e/framework/events/events.go]: BUILD file is missing a line for this
```
compilepkg: missing strict dependencies:
/b/f/w/test/e2e/framework/events/events.go: import of "k8s.io/apimachinery/pkg/types"
```

Comment (BobyMCbobs) [reply test/e2e/framework/events/events.go]: Fixed in ac085fb

**Comment by hh** [comment]
@BobyMCbobs please rebase and retest.

**Comment by hh** [comment]
@spiffxp This is OK, we are not relying on certain things to happen... this is testing the event infrastructure NOT the side effects of other behavior. Could be subject to delivery of events isn't certain, but seems ok.
@smarterclayton we should test the API, and the reporter. Clients using the . if you can't send a valid event to the informer. In practice if the event sending infra can't succeed. The test should test the basic crud and verify, this should not hit those edge cases.

For the scope of this PR @spiffxp will review and it will go through.

File: test/e2e/framework/events/events.go
Comment (spiffxp) [inline test/e2e/framework/events/events.go]:
```suggestion
		framework.ExpectEqual(foundCreatedEvent, false, "should not have found test event after deletion")
```

**Review by spiffxp — CHANGES_REQUESTED** [review]
One nit, otherwise LGTM

**Comment by spiffxp** [comment]
/priority important-soon
/approve
Will /lgtm when nit is addressed

**Pull request merged**
